### PR TITLE
Update OpenShift Template - various changes

### PR DESCRIPTION
**Pull request by rbo**
Hi folks,

after deploying & playing  with Skydive on OpenShift Container Platform and ran into Issue #1391 here some improvements for skydive on OpenShift:

 - Change installation process to use git URL's instead of files. You don't have to clone the repo for an installation.
 - OpenShift provides per default SSL encryption, so change the route from HTTP to HTTPS
 - Add k8s analyser probe - inspired from pull request #1419 with some cleanups
 - Cleanup Skydive configuration, idea is use only environment variables on one place for analyser and agent.
     - Removed the configmap, from my point of view config map make only sense if you share the same configurations to many other objects, with skydive that is not the case. Analyser and agent has it own config map.
    - Additional for the agend deploymentconfig you some benefits to add the environment variables direct to the deploymentconfig:
        - If you change the config, the configchangetrigger redeploy
   skydive-analyzer with the new configuration
        - The configuration is versioned between different deployments
        - You can easily rollback to an old configuration
 - Parameterize OpenShift template - add the possibility  to set loglevel add installation process
 - Cleanup README

Let's discuss the changes, hope it helps to improve the skydive installation on openshift.

Thanks
Robert

### CODE REVIEW

**Comment by safchain** [comment]
@SchSeba can you please have a look at this ?

File: contrib/openshift/skydive-template.yaml [hunk 1]
```diff
@@ -167,9 +156,12 @@ objects:
...
             value: $(SKYDIVE_ANALYZER_SERVICE_HOST):$(SKYDIVE_ANALYZER_SERVICE_PORT_API)
-          envFrom:
-          - configMapRef:
-              name: skydive-agent-config
+          - name: SKYDIVE_LOGGING_LEVEL
+            value: ${SKYDIVE_LOGGING_LEVEL}
+          - name: SKYDIVE_AGENT_TOPOLOGY_PROBES
+            value: "ovsdb runc docker"
+          - name: SKYDIVE_RUNC_PATH
+            value: "/run/containerd/runc /var/run/runc /var/run/runc-ctrs"
```
Comment (SchSeba) [inline contrib/openshift/skydive-template.yaml]: The `/run/containerd/runc` is not mapped

File: contrib/openshift/skydive-template.yaml [hunk 2]
```diff
@@ -220,8 +212,16 @@ objects:
...
       kind: Service
       name: skydive-analyzer
       weight: 100
     wildcardPolicy: None
+parameters:
+- description: Loglevel of Skydive agent and analyzer
+  displayName: Loglevel
+  name: SKYDIVE_LOGGING_LEVEL
+  required: true
+  value: INFO
```
Comment (SchSeba) [inline contrib/openshift/skydive-template.yaml]: add  a new line please

File: contrib/openshift/README.md [hunk 3]
```diff
@@ -1,50 +1,49 @@
...
 
-Check that everything is working and created:
+####  Install from OpenShift template
 
 ```
-oc get pods
-oc get ds
+# adjust VERSION for the current version - for example: v0.20.1 or master
+VERSION=v0.20.1
+oc process -f https://raw.githubusercontent.com/skydive-project/skydive/${VERSION}/contrib/openshift/skydive-template.yaml | oc apply -f -
```
Comment (SchSeba) [inline contrib/openshift/README.md]: @safchain if we want to point the yaml this way you will need to create a new tag.

File: contrib/openshift/README.md [hunk 4]
```diff
@@ -1,50 +1,49 @@
...
-```
 
-Check that everything is working and created:
+####  Install from OpenShift template
 
 ```
-oc get pods
-oc get ds
+# adjust VERSION for the current version - for example: v0.20.1 or master
+VERSION=v0.20.1
```
[the remark this answers is not in the document]
Comment (rbo) [reply contrib/openshift/README.md]: Changed.

File: contrib/openshift/README.md [hunk 5]
```diff
@@ -1,50 +1,49 @@
...
-Expose your route:
+#### Check that everything is working and created:
 
-```
-oc delete route skydive-analyzer
-oc expose svc skydive-analyzer
-```
+ - Overall status: `oc status`
+ - List all pods: `oc get pods`
+ - List all daemonsets: `oc get daemonset`
```
[the remark this answers is not in the document]
Comment (rbo) [reply contrib/openshift/README.md]: Changed.

File: contrib/openshift/skydive-template.yaml (hunk 1 above)
Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: Indeed, thangs. Changed.

File: contrib/openshift/skydive-template.yaml (hunk 2 above)
Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: Changed with commit ```Add missing hostpath mount /run/containerd/runc```

File: contrib/openshift/README.md (hunk 3 above)
Comment (rbo) [reply contrib/openshift/README.md]: Or use ```VERSION=master```

**Comment by rbo** [comment]
Changed everything. Thanks for input!

File: contrib/openshift/README.md (hunk 3 above)
Comment (safchain) [reply contrib/openshift/README.md]: So let's use master for now because I don't know when we will release a new version.

Comment (rbo) [reply contrib/openshift/README.md]: It means we can remove the VERSION completely? 
```suggestion
oc process -f https://raw.githubusercontent.com/skydive-project/skydive/master/contrib/openshift/skydive-template.yaml | oc apply -f -
```

File: contrib/openshift/README.md [hunk 6]
```diff
@@ -1,50 +1,50 @@
...
-Install the template
+#### Create a new project with an empty node selector to asume that Skydive runs on all nodes. 
 
 ```
-oc create -f skydive-template.yaml
+oc adm new-project --node-selector='' skydive
 ```
 
-You need the DeploymentConfig to run in privileged:
+####  Skype analyser and agent need  extended  privileges
```
Comment (lebauce) [inline contrib/openshift/README.md]: s/analyser/analyzer/

File: contrib/openshift/README.md [hunk 7]
```diff
@@ -1,50 +1,50 @@
...
 
 ```
-oc adm policy add-scc-to-user privileged system:serviceaccount:skydive:default
+# analyser and agen run as privileged container
```
Comment (lebauce) [inline contrib/openshift/README.md]: s/analyser/analyzer/
s/agen/agent/

File: contrib/openshift/README.md [hunk 8]
```diff
@@ -1,50 +1,50 @@
...
+oc adm policy add-scc-to-user privileged -z default
+# analyser need cluster-reader access get all informations from the cluster
```
Comment (lebauce) [inline contrib/openshift/README.md]: s/analyser/analyzer/

File: contrib/openshift/skydive-template.yaml [hunk 9]
```diff
@@ -82,9 +63,17 @@ objects:
...
           - --listen=0.0.0.0:8082
-          envFrom:
-          - configMapRef:
-              name: skydive-analyzer-config
+          env:
+          - name: SKYDIVE_ANALYZER_FLOW_BACKEND
+            value: elasticsearch
+          - name: SKYDIVE_ANALYZER_TOPOLOGY_BACKEND
+            value: elasticsearch
+          - name: SKYDIVE_ANALYZER_TOPOLOGY_PROBES
```
Comment (lebauce) [inline contrib/openshift/skydive-template.yaml]: Until now, we did not enable the k8s probe by default as it has some scalability issues.
A mitigation would be to limit the types of Kubernetes resources we are interested in through the "probes" section of the k8s probe : https://github.com/skydive-project/skydive/blob/master/etc/skydive.yml.default#L129
For example, only enable "cluster", "service", "pod" and "node".

File: contrib/openshift/README.md (hunk 6 above)
Comment (rbo) [reply contrib/openshift/README.md]: Thanks, changed.

File: contrib/openshift/README.md (hunk 7 above)
Comment (rbo) [reply contrib/openshift/README.md]: Thanks, changed.

File: contrib/openshift/skydive-template.yaml (hunk 9 above)
Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: Got it, what do you think about:
 - Leave `SKYDIVE_ANALYZER_TOPOLOGY_PROBES` empty by default.
 - Add a template variable for `SKYDIVE_ANALYZER_TOPOLOGY_PROBES` so if like you can add `k8s` at deployment time.
 - And a second template variable for k8s probes. 
   @lebauce do you know the name of the env. var? 
   Should be `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES="cluster service pod node"`
   Default also empty.

File: contrib/openshift/skydive-template.yaml [hunk 10]
```diff
@@ -201,13 +195,16 @@ objects:
...
         - hostPath:
-            path: /run/runc
+            path: /var/run/runc
           name: runc
         - hostPath:
-            path: /run/runc-ctrs
+            path: /var/run/runc-ctrs
           name: runc-ctrs
+        - hostPath:
+            path: /run/containerd/runc
```
[the remark this answers is not in the document]
Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: A good questions. In general I never thought about `/run` and `/var/run/`.  

What I learned: In FHS 3.0, /var/run is replaced by /run; a system should either continue to provide a /var/run directory, or provide a symbolic link from /var/run to /run, for backwards compatibility. [1]

In my OpenShift / RHEL 7.6 System:
```sh
$ stat /run
  File: ‘/run’

$ stat /var/run
  File: ‘/var/run’ -> ‘../run’

```

It means to me: I will change everything to `/run`

Thanks
Robert
[1] https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s13.html

File: contrib/openshift/README.md (hunk 3 above)
Comment (rbo) [reply contrib/openshift/README.md]: Thanks, changed.

File: contrib/openshift/skydive-template.yaml (hunk 9 above)
Comment (lebauce) [reply contrib/openshift/skydive-template.yaml]: Sounds perfect. Yes, it should be `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES

Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: If I tried to start the analyzer with `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES` than I got an `SIGSEGV` error:
```
2019-01-17T14:26:17.666Z	INFO	k8s/probe.go:151 NewProbe	skydive-analyzer-1-r75v5: Probe k8s subprobes [cluster service pod node]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xd19aca]

goroutine 1 [running]:
github.com/skydive-project/skydive/graffiti/graph.(*Indexer).Start(0xc4209349c0)
	/root/go/src/github.com/skydive-project/skydive/graffiti/graph/indexer.go:161 +0x2a
github.com/skydive-project/skydive/topology/probes/k8s.newNamespaceLinker(0xc4206f93e0, 0x21aeb36, 0x3, 0xc420589c80, 0xc, 0xc, 0x432268, 0xc0)
	/root/go/src/github.com/skydive-project/skydive/topology/probes/k8s/namespace.go:60 +0x209
github.com/skydive-project/skydive/topology/probes/k8s.(*Probe).AppendNamespaceLinkers(0xc42088ed00, 0xc420589c80, 0xc, 0xc)
	/root/go/src/github.com/skydive-project/skydive/topology/probes/k8s/probe.go:140 +0x65
github.com/skydive-project/skydive/topology/probes/k8s.NewK8sProbe(0xc4206f93e0, 0xc42088aed0, 0xc4200540b1, 0x3)
	/root/go/src/github.com/skydive-project/skydive/topology/probes/k8s/k8s.go:117 +0xa5d
github.com/skydive-project/skydive/analyzer.NewTopologyProbeBundleFromConfig(0xc4206f93e0, 0xc4201a8a30, 0xc4206f93e0, 0xc420866f60)
	/root/go/src/github.com/skydive-project/skydive/analyzer/probes.go:53 +0x4e6
github.com/skydive-project/skydive/analyzer.NewServerFromConfig(0xc420219fe0, 0x21f31f7, 0x1f)
	/root/go/src/github.com/skydive-project/skydive/analyzer/server.go:282 +0xd5a
github.com/skydive-project/skydive/cmd/analyzer.glob..func1(0x35a0380, 0xc4203b4b10, 0x0, 0x3)
	/root/go/src/github.com/skydive-project/skydive/cmd/analyzer/analyzer.go:49 +0xe8
github.com/skydive-project/skydive/vendor/github.com/spf13/cobra.(*Command).execute(0x35a0380, 0xc4204b6ac0, 0x3, 0x4, 0x35a0380, 0xc4204b6ac0)
	/root/go/src/github.com/skydive-project/skydive/vendor/github.com/spf13/cobra/command.go:603 +0x234
github.com/skydive-project/skydive/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x35a5aa0, 0x0, 0x359ff40, 0xc42062ff58)
	/root/go/src/github.com/skydive-project/skydive/vendor/github.com/spf13/cobra/command.go:689 +0x2d4
github.com/skydive-project/skydive/vendor/github.com/spf13/cobra.(*Command).Execute(0x35a5aa0, 0xc42062ff78, 0xc4200b8058)
	/root/go/src/github.com/skydive-project/skydive/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.main()
	/root/go/src/github.com/skydive-project/skydive/skydive.go:32 +0x2d
```
I tried with
 - `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES="cluster"` 
 - `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES="node"`
 - `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES="cluster service pod node"`

all got the same error.

Analyzer version:`Skydive Analyzer 0.21.0-f88335bd44dc`

@lebauce do you have some informations about that?

File: contrib/openshift/skydive-template.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: Resolved.

File: contrib/openshift/skydive-template.yaml (hunk 9 above)
Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: Added the changes, wait of feedback from @lebauce. If you leave `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES` empty all works very well.

Comment (lebauce) [reply contrib/openshift/skydive-template.yaml]: Interesting. This is a probably a bug in the `k8s` probe (cc @hunchback). Can you try with SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES="namespace cluster service pod node". If it works, I think we can merge your PR as it

Comment (rbo) [reply contrib/openshift/skydive-template.yaml]: With `SKYDIVE_ANALYZER_TOPOLOGY_K8S_PROBES="namespace cluster service pod node"` it works very well.

**Pull request merged**